src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -1246,17 +1223,58 @@ struct BlockZipperJoinImpl {
     }
   }
 
+  // Consume the blocks until the first block is found that does contain a
+  // defined value. All blocks up until that point are stored in
+  // `side.undefBlocks_` and skipped for subsequent processing.
+  void findFirstBlockWithoutUndef(auto& side) {
+    auto& it = side.it_;
+    const auto& end = side.end_;
+    while (it != end) {
+      auto& el = *it;
+      if (!std::ranges::empty(el) && isUndefined_(el.front())) {
+        bool endUndefined = isUndefined_(el.back());
+        side.undefBlocks_.emplace_back(std::move(el));
+        ++it;
+        if (!endUndefined) {
+          auto& lastUndefinedBlock = side.undefBlocks_.back();
+          side.currentBlocks_.push_back(lastUndefinedBlock);
+          auto subrange = std::ranges::equal_range(
+              lastUndefinedBlock.subrange(),
+              lastUndefinedBlock.subrange().front(), lessThan_);
+          size_t undefCount = std::ranges::size(subrange);
+          lastUndefinedBlock.setSubrange(std::move(subrange));
+          auto& firstDefinedBlock = side.currentBlocks_.back();
+          firstDefinedBlock.setSubrange(
+              firstDefinedBlock.fullBlock().begin() + undefCount,
+              firstDefinedBlock.fullBlock().end());
+          break;
+        }
+        continue;
+      }
+      break;
+    }
+  }
+
+  // Find and process all leading undefined values from the blocks.
+  void fetchAndProcessUndefinedBlocks() {
+    if constexpr (potentiallyHasUndef) {
+      findFirstBlockWithoutUndef(leftSide_);
+      findFirstBlockWithoutUndef(rightSide_);
+      addCartesianProduct(leftSide_.undefBlocks_, rightSide_.undefBlocks_);
+    }
+  }
+
+  // The actual join routine that combines all the previous functions.
   template <bool DoOptionalJoin>
   void runJoin() {
+    fetchAndProcessUndefinedBlocks();
```

Comment:
How hard would it be, 
to detect after this call that we have `potentiallyUndef == true`, but actually no undef blocks in any input, and to then change into the statically `potentiallyUndef == false` case, for better code generation? 
From afar we could either put the template on all the functions for which it is relevant (so the `potentiallyUndef` is not a parmeter of the class anymore). 
Or you transparently convert after detecting to an instance of the template, where `potentiallyUndef` is `false` (by moving all the state, and hoping/testing that everything works as expected).
The first one would probably be cleaner and also feasible.